### ISSUE TITLE
redirect_uri is missing but it seems we have it all configured correctly

### ISSUE THREAD

**Issue by chaseklingelzen** [plaid/plaid-link-ios#30]
---
### A detailed description of the steps to reproduce the issue

When selecting Platpus OAuth Bank (or any bank for that matter), we see the following warning banner:

![File (27)](https://github.com/plaid/plaid-link-ios/assets/94926620/efb7ab84-995e-4b93-a0c1-a3347713b377)

We have configured our redirect uri in the Plaid dashboard.

<img width="652" alt="Screenshot 2024-01-08 at 3 40 49 PM" src="https://github.com/plaid/plaid-link-ios/assets/94926620/cbdc35da-18b7-4a8c-b15e-49e9a95994ee">

**Assume that our uri is the following:**

**https://mycompany.com/n/native-oauth-plaid-link-completed**

in Xcode, we have set up applinks like so:

<img width="1006" alt="Screenshot 2024-01-08 at 3 24 58 PM" src="https://github.com/plaid/plaid-link-ios/assets/94926620/5990ef7c-5672-421f-a25d-af982691cb86">

<img width="1107" alt="Screenshot 2024-01-08 at 3 25 17 PM" src="https://github.com/plaid/plaid-link-ios/assets/94926620/8f2f3e30-e197-450e-b957-d298b65da77a">

When we create the link token we pass the redirectUrl "https://mycompany.com/n/native-oauth-plaid-link-completed".

We have an apple-app-site-association file being hosted. Here is the basic outline of our apple-app-site-association:

<img width="670" alt="Screenshot 2024-01-08 at 3 27 44 PM" src="https://github.com/plaid/plaid-link-ios/assets/94926620/faeb6c61-01a0-4c1c-84ad-56c548bea5b6">

### What you expected to see?

I expected that our redirect uri was set up correctly and the banner in the screenshot above stating we have a missing redirect_uri is not present.

### What you saw instead?

The banner saying we have potential issues with our redirect_uri.

---

When submitting an issue for Link on the web or within a Webview, please include the following information:

N/A

When submitting an issue for [**Plaid Link iOS**](https://github.com/plaid/link/tree/master/ios) please include the following information:

| Name            | Value |
|-----------------|-------|
| Link `env`      |   Sandbox    |
| LinkKit version |  5.0.0     |
| iOS version     |   17.1.2    |
| iOS device      |   iPhone 15 pro    |

Note: This happens on both physical device and simulators of various versions and device types. 

#### List of used 3rd party libraries (e.g. CocoaPods):

* [Plaid](https://cocoapods.org/pods/Plaid)
*
*

**Comment by ToddKerpelman** [comment]
That certainly looks right to me -- can you share the code you're using to generate the Link token?

**Comment by chaseklingelzen** [comment]
Hi Todd, trying to determine what code would be most helpful to share. I assume you'd want to see the server side code that generates the link token? 

I think the main questions I'm looking to get answered here is the following:

1. Does this warning mean we will run into issues in production? If so, can you outline them?
2. Is it possible to test app2app flows on mobile? I found [this](https://plaid.com/docs/link/oauth/) documentation stating the following:

_You can test OAuth on Sandbox even if Plaid has not yet enabled OAuth flows for your account. To test out the OAuth flow in the Sandbox environment, you can use Platypus OAuth Bank (ins_127287), Platypus OAuth App2App Bank (ins_132241) for testing App2App flows on mobile, or to emulate European flows, Flexible Platypus Open Banking (ins_117181). These dummy institutions direct you to a Plaid sample OAuth flow that is similar to what you would see for a bank’s OAuth flow. When prompted, you can enter anything as credentials (including leaving the input fields blank) to proceed through the sample OAuth flow. Note that institution-specific OAuth flows cannot be tested in Sandbox; OAuth panes for Platypus institutions will be shown instead._ 

But then found [this](https://plaid.com/docs/link/oauth/#app-to-app-authentication
) documentation, which seems to state you can't actually test app2app flows in sandbox:

_If using webviews, App-to-App support is not automatic; for an App-to-App experience, it is strongly recommended to use a Plaid mobile SDK instead.
The full App-to-App flow cannot currently be tested in the Sandbox environment. In Sandbox, all OAuth flows will appear as a webview on top of your current application. It is recommended to test iOS App-to-App flows on physical devices, as some customers have reported issues with testing Universal Links flows on iOS simulators.
Chase is currently the only bank that supports App-to-App authentication on Plaid._

So, any clarification here would be greatly appreciated!

**Comment by ToddKerpelman** [comment]
* Yep -- the server side code that generates the Link Token. (Specifically, I'm wondering if the redirect_uri is in the wrong place or there's something else funny with it) 
* The warning does basically mean that the app2app flow won't work in production, assuming you're building a native iOS app. This primarily affects Chase at the moment, but supporting it will help conversion there. 
* At the moment, the app2app flow cannot be tested on sandbox. The team is looking to add support for it soon, but in the meantime I'll go ahead and update that first link to make that clearer.

**Comment by chaseklingelzen** [comment]
I'll have to double check with my team that I can share the server side code here. Will get back to you on that tomorrow. Thanks @ToddKerpelman. 

So, if we do not see the redirect_uri banner any longer, we can safely assume the app2app flow is working? Is that the only sort of "signifier" that things are working in sandbox?

**Comment by chaseklingelzen** [comment]
Hi @ToddKerpelman, please let me know if this snippet gives you enough information:

```Python
def create_link_token(
    account_uuid: str,
    account_filters: Dict[str, Dict[str, List[str]]],
    link_customization_name: str,
    redirect_url: str = None, // note it's my understanding that we are passing a value here.
    products: List = None,
    plaid_access_token: str = None,
    account_selection_enabled: bool = False,
    android_package_name: str = None,
) -> Dict[str, str]:
    """
    Creates a link token for the given account id and returns the link token

    @param account_uuid: the account id for which the link token is being created
    @param account_filters: the account filters for the link token
    @param link_customization_name: the name of the link customization
    @param redirect_url: the redirect url for the link token
    @param products: the products for the link token
    @param plaid_access_token: the plaid access token for the link token
    @param account_selection_enabled: whether the account selection is enabled
    @param android_package_name: the android package name

    @return: the link token
    ...
        conditional_args = {}
    ...
        # The name of your app's Android package. Required if using the `link_token`
        # to initialize Link on Android. Any package name specified here must also
        # be added to the Allowed Android package names setting on the
        # [developer dashboard](https://dashboard.plaid.com/team/api).
        # When creating a `link_token` for initializing Link on other platforms,
        # `android_package_name` must be left blank and `redirect_uri`
        # should be used instead.
        if android_package_name:
            conditional_args["android_package_name"] = android_package_name
      elif redirect_url:
          conditional_args["redirect_uri"] = redirect_url
     plaid_request = LinkTokenCreateRequest(
         client_name="MyCompany",
         country_codes=[CountryCode("US")],
         language="en",
         user=LinkTokenCreateRequestUser(client_user_id=account_uuid),
         webhook=settings.PLAID_WEBHOOKS_URL,
         **conditional_args,
   )

token_response = client.link_token_create(plaid_request)
```

**Comment by ToddKerpelman** [comment]
That looks good, although there's no chance you're passing in both the Android package name and the redirect_url in the same call, is there? (Because if you include the first, it won't set the latter.) If that looks okay, I might try logging the values of your conditional args to make sure those really are getting set properly.

And if that still looks like your setting your values okay, I might file a ticket with support. They might be able to find something that I can't.

**Comment by ToddKerpelman** [comment]
> So, if we do not see the redirect_uri banner any longer, we can safely assume the app2app flow is working? Is that the only sort of "signifier" that things are working in sandbox?

Actually, the bigger thing to check would be that Universal Links are working, and that opening up that URL on an iOS device (or the simulator) actually opens up your app. I go into details on how to test that [here](https://youtu.be/9fgmW38usTo?si=l_0-JC7vDAiVIljK&t=3354)

**Comment by chaseklingelzen** [comment]
> That looks good, although there's no chance you're passing in both the Android package name and the redirect_url in the same call, is there? (Because if you include the first, it won't set the latter.) If that looks okay, I might try logging the values of your conditional args to make sure those really are getting set properly.
> 
> And if that still looks like your setting your values okay, I might file a ticket with support. They might be able to find something that I can't.

My understanding is we should only be sending the redirect_url if the client is iOS, but we will triple check that. Sounds good on filing a ticket if we continue to hid a roadblock there.

**Comment by chaseklingelzen** [comment]
> > So, if we do not see the redirect_uri banner any longer, we can safely assume the app2app flow is working? Is that the only sort of "signifier" that things are working in sandbox?
> 
> Actually, the bigger thing to check would be that Universal Links are working, and that opening up that URL on an iOS device (or the simulator) actually opens up your app. I go into details on how to test that [here](https://youtu.be/9fgmW38usTo?si=l_0-JC7vDAiVIljK&t=3354)

Makes sense on this! I did see that in the video you posted. I was doing some testing around this today actually. I did have one question here:

Should we be removing "www" from our redirect uri in the plaid dashboard or does it not really matter (e.g. "https://www.mycompany.com/n/native-oauth-plaid-link-completed" vs. "https://www.mycompany.com/n/native-oauth-plaid-link-completed")

**Comment by chaseklingelzen** [comment]
Hi @ToddKerpelman, 

Wanted to follow up here. The root cause of our issue was actually on our server side logic. We were not passing the redirect_url properly to that python method I mentioned. 

For anyone else who finds this thread, it was important that the redirect url we defined match exactly what we sent to generate the link token. 

redirect uri defined in plaid: "https://www.mycompany.com"
redirect uri sent from iOS client to server ultimately to plaid must be: "https://www.mycompany.com" 

If you omit the "www", it will not work. 

Side note @ToddKerpelman - I built my first ever iOS app with Firebase and utilized a lot of the videos you created to make that happen. You're an incredible teacher and I really appreciate the content you've created over the year. Plaid is in good hands!
